### PR TITLE
feat: support macos by sharing memory on ordinary file

### PR DESCRIPTION
**Pull request by Breeze-P**
As the title said.

### CODE REVIEW

File: src/memfd.rs [hunk 1]
```diff
@@ -20,21 +20,59 @@ pub struct MemfdSettings {
...
 
-    let file = unsafe { fs::File::from_raw_fd(raw_fd) };
-    file.set_len(size).context(error::IoSnafu)?;
+        Ok(file)
+    }
+    #[cfg(target_os = "macos")]
+    {
+        let MemfdSettings { name, size } = settings;
+
+        let base_path = String::from("/tmp/shm/");
```
Comment (fengys1996) [inline src/memfd.rs:47]: There seems to be no cleanup mechanism after normal files are created, which may result in more and more files being created.

Comment (fengys1996) [reply src/memfd.rs:47]: The file path doesn't seem to be configurable either.

Comment (Breeze-P) [reply src/memfd.rs]: The new commit solved that.

**Comment by fengys1996** [comment]
I suggest
1. Add path field in MemfdSettings.
```rust
/// Settings for creating a memfd.
#[derive(Debug, Clone)]
pub struct MemfdSettings {
    /// The name of the memfd. Only used for debugging.
    pub name: String,
    /// The size of the memfd.
    pub size: u64,
    /// The path of the memfd.
    #[cfg(target_os = "macos")]
    pub path: PathBuf,
}
```
2. And should change name `MemfdSettings` -> `Settings`, `memfd_create` -> `create_fd`.
3.
```rust
#[derive(Debug, Clone)]
pub struct ProducerSettings {
    pub(super) grpc_sock_path: PathBuf,
    pub(super) fdpass_sock_path: PathBuf,
    pub(super) ringbuf_len: usize,
    pub(super) heartbeat_interval: Duration,
    pub(super) result_fetch_retry_interval: Duration,
    #[cfg(target_os = "macos")]
    pub(super) backed_file_path: Duration,
}
```

**Comment by fengys1996** [comment]
Any update? @Breeze-P

File: src/fd_pass.rs [hunk 2]
```diff
@@ -273,7 +274,7 @@ mod tests {
             let path_c = path.clone();
             let join = tokio::spawn(async move {
                 let file = tempfile::tempfile().unwrap();
-                file.set_len(min_ringbuf_len()).unwrap();
+                file.set_len(page_align_size(min_ringbuf_len())).unwrap();
```
Comment (fengys1996) [inline src/fd_pass.rs:277]:
```suggestion
                file.set_len(min_ringbuf_len()).unwrap();
```

File: src/memfd.rs [hunk 3]
```diff
@@ -1,49 +1,102 @@
...
+        // In order to reduce the writing to the disk, we can delete the file immediately, why?
+        // Refer to: https://stackoverflow.com/questions/39779517/does-mac-os-have-a-way-to-create-an-anonymous-file-mapping
+        // But we are only developing on macOS and will not use macOS as a production environment. So we will comment out the following.
+        fs::remove_file(path).context(error::IoSnafu)?;
+
+        let raw_fd = owned_fd.into_raw_fd();
+
+        let file = unsafe { fs::File::from_raw_fd(raw_fd) };
+        file.set_len(size).context(error::IoSnafu)?;
+
```
Comment (fengys1996) [inline src/memfd.rs:90]: disable_shrink_or_grow(raw_fd)?;

File: src/memfd.rs [hunk 4]
```diff
@@ -1,49 +1,102 @@
...
+        Ok(file)
+    }
+    #[cfg(not(any(
+        target_os = "linux",
+        target_os = "android",
+        target_os = "freebsd"
+    )))]
+    {
+        let Settings { name, size, path } = settings;
+        let _ = name;
```
Comment (fengys1996) [inline src/memfd.rs:64]:
```suggestion
        let Settings { name: _, size, path } = settings;
```

File: src/producer/settings.rs [hunk 5]
```diff
@@ -81,13 +104,41 @@ impl ProducerSettingsBuilder {
...
 
-        ProducerSettings {
+        #[cfg(not(any(
+            target_os = "linux",
+            target_os = "android",
+            target_os = "freebsd"
+        )))]
+        let backed_file_path = self
+            .backed_file_path
+            .unwrap_or_else(|| PathBuf::from("/tmp/shm"));
```
Comment (fengys1996) [inline src/producer/settings.rs:114]:
```suggestion
            .unwrap_or_else(|| PathBuf::from("/tmp/shm.sock"));
```

**Review by fengys1996 — APPROVED** [review]
Good Job!

**Pull request merged**